### PR TITLE
Add some text about env vars when linking containers

### LINKED ISSUE #5169: envvars which are enherited from a linked container can cause a security issue
Many times we use env vars to do things like give credentials for db access, etc.
Right now a container which links to other containers is pulling in the env from the linked container, thus exposing these credentials to containers that shouldn't really have them.

### PR DESCRIPTION
**Pull request by duglin**
In particular, the security implications.

Closes #5169

Signed-off-by: Doug Davis dug@us.ibm.com

### CODE REVIEW

File: docs/sources/userguide/dockerlinks.md [hunk 1]
```diff
@@ -206,12 +206,30 @@ from the URL (e.g. `WEBDB_PORT_8080_TCP_PROTO=tcp`).
...
 
-Finally, there will be an environment variable called `<alias>_PORT` that will
+Additionally, there will be an environment variable called `<alias>_PORT` 
+that will
 contain the URL of the first exposed port of the source container.
 For example, `WEBDB_PORT=tcp://172.17.0.82:8080`. In this case, 'first'
 is defined as the lowest numbered port that is exposed. If that port is
 used for both tcp and udp, then the tcp one will be specified.
 
+Finally, Docker will expose each "Docker managed" environment variable
```
Comment (moxiegirl) [inline docs/sources/userguide/dockerlinks.md]: Doug this page contains a lot of structural issues and this section in particular. 

* It is mixing tenses within sentences
* multiple use of future tense (not good practice in tech docs)
* over use of notes indicating structural problems in the text
* confused presentation of variable formats

All this in one section makes it _very_ hard to parse. Not your fault, but as you are knowledgeable and in the file, maybe you could extend your change a bit and review what I pass you? I'd really appreciat it -- I know linking is an important part of the product.  

----

### Environment Variables

Docker creates several environment variables when you link containers. Docker   automatically creates environment variables in the target container based on the `--link` parameters. Others environment variables Docker sets in the target are 
Docker _managed environment variables_. Managed environment variables are are generated from two places:

* the `ENV` commands in the source container's Dockerfile
* through the `-e` options on the `docker run` command when the source container is started

These environment variables enable programmatic discovery from within the target container of information related to the source container.

> **Warning**:
> It is important to understand that *all* environment variables managed
> by Docker within a container are made available to *any* container 
> that links to it. This could have serious security implications if sensitive
> data is stored in them.

Docker sets an `<alias>_NAME` environment variable for each target container listed in the `--link` parameter. For example, if a new container called `web` is linked to a database container called `db` via `--link db:webdb`,  then Docker creates a `WEBDB_NAME=/web/webdb` variable in the `web` container.

Docker also defines a set of environment variables for each port exposed by the
source container.  Each variable has a unique prefix in the form:

`<name>_PORT_<port>_<protocol>`

The components in this prefix are:
 
* the alias <name> specified in the `--link` parameter (for example, `webdb`)
* the <port> number exposed
* a <protocol> which is either TCP or UDP

Docker uses this prefix format to define three distinct environment variables:

* The `prefix_ADDR` variable contains the IP Address from the URL, for example `WEBDB_PORT_8080_TCP_ADDR=172.17.0.82`.
* The `prefix_PORT` variable contains just the port number from the URL for example `WEBDB_PORT_8080_TCP_PORT=8080`.
* The `prefix_PROTO` variable contains just the protocol from the URL for example 1WEBDB_PORT_8080_TCP_PROTO=tcp1`.

If the container exposes multiple ports, an environment variable set is defined for each one. This means, for example, if a container exposes 4 ports that Docker creates 12 environment variables, 3 for each port.

Additionally, Docker creates an environment variable called `<alias>_PORT`. This variable contains the URL of the source container's first exposed port. The  'first' port is defined as the exposed port with the lowest number.
For example, consider the `WEBDB_PORT=tcp://172.17.0.82:8080` variable.  If that port is used for both tcp and udp, then the tcp one is specified.

Finally, Docker also exposes each Docker managed environment variable
from the source container as an environment variable in the target. For each `ENV` command in the Dockerfile or option passed with `docker run -e`, Docker creates an `<alias>_ENV_<name>` variable in the target container. The variable's value is set to the value Docker used when it started the source.

Returning back to our database example, you can run the `env`
command to list the specified container's environment variables.

```
    $ sudo docker run --rm --name web2 --link db:db training/webapp env
    . . .
    DB_NAME=/web2/db
    DB_PORT=tcp://172.17.0.5:5432
    DB_PORT_5432_TCP=tcp://172.17.0.5:5432
    DB_PORT_5432_TCP_PROTO=tcp
    DB_PORT_5432_TCP_PORT=5432
    DB_PORT_5432_TCP_ADDR=172.17.0.5
    . . .
```

You can see that Docker has created a series of environment variables with
useful information about the source `db` container. Each variable is prefixed with
`DB_`, which is populated from the `alias` you specified above. If the `alias`
were `db1`, the variables would be prefixed with `DB1_`. You can use these
environment variables to configure your applications to connect to the database
on the `db` container. The connection will be secure and private; only the
linked `web` container will be able to talk to the `db` container.

#### Important notes on Docker environment variables

Unlike host entries in the [`/etc/hosts` file](#updating-the-etchosts-file),
IP addresses stored in the environment variables are not automatically updated
if the source container is restarted. We recommend using the host entries in
`/etc/hosts` to resolve the IP address of linked containers.

These environment variables are only set for the first process in the container. Some daemons, such as `sshd`, will scrub them when spawning shells for connection.

Comment (estesp) [inline docs/sources/userguide/dockerlinks.md]: I know it's hard to know what to call them, but "Docker managed" might be replaced with something like: *Docker will expose all environment variables originating from Docker from the source container..*

File: docs/sources/userguide/dockerlinks.md [hunk 2]
```diff
@@ -206,12 +206,30 @@ from the URL (e.g. `WEBDB_PORT_8080_TCP_PROTO=tcp`).
...
+from the source container as a new environment variable in the target
+container. 
+Docker managed environment variables are the ones defined by any `ENV`
+command in the Dockerfile used to create the source container, or any defined
+using the `-e` option on the `docker run` command when the source container
+was started.
```
Comment (estesp) [inline docs/sources/userguide/dockerlinks.md]: If you change it above, this could become: *Environment variables originating from Docker include both any `ENV` commands in the Dockerfile used to create the source source container and those defined with the `-e` option on the `docker run` command..*

**Comment by moxiegirl** [comment]
Doug, 

I'm gonna push your patience and ask if you coudl tweak some of the headings while you are in the file. You are a unique combination of knowledgeable and fast --- if you look at the headings in the existing file...it makes the content hard to navigate.  I've looked at the content and tweaked the levels and titles.  I think what I have is easier to follow:

image


I'll send you the markdown for this.  If you are feeling generous or whatever and want to do it too...that would be great. I totally get it if you don't want to...I'll create a new PR and drop it in myself.

**Comment by duglin** [comment]
@moxiegirl @estesp I tried to take into account both of your comments/edits - see what you think

**Comment by moxiegirl** [comment]
LGTM

**Comment by thaJeztah** [comment]
Currently commuting so haven't read extensively, but are port-*ranges* also describes in the section on environment variables? (i.e. `-p 1000-2000`)

**Comment by duglin** [comment]
Nope this section doesn’t mentions ranges

**Comment by crosbymichael** [comment]
LGTM
